### PR TITLE
[CPU][Refactoring] Introduce VariableExecutor

### PR DESCRIPTION
**Pull request by EgorDuplensky**
Depending on the parameters a `FullyConnected` node can
use one or multiple executors.
With the current approach, even when just a single executor
is used, every `prepareParams()` (executor::update())
call goes through executor selection routine.

The idea is to avoid such `prepareParams()` overhead for a single executor scenarious,
which are probably the most common ones.

Thus, split the pipeline input two branches:
- only single simple executor is used and updated
- a `VariableExecutor` is used and updated. `VariableExecutor` contains
  two or more simple executors

### CODE REVIEW

**Comment by EgorDuplensky** [comment]
@dmitry-gorokhov Could you please take a look?

File: src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp [hunk 1]
```diff
@@ -284,11 +162,11 @@ class ExecutorFactory {
     std::vector<ExecutorPtr> m_executors;
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp:162]: Seems like `m_implementationRequiresFallback` and `m_executors` fields should be removed

File: src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp [hunk 2]
```diff
@@ -95,104 +67,41 @@ class ExecutorFactory {
...
-        (void)create(implId, memory, m_context);
-    }
-
-    /**
-     * @brief Creates an Executor instance based on provided memory arguments.
-     *
-     * Creates an Executor instance using the provided MemoryArgs, selecting an appropriate implementation
-     * based on the characteristics of the memory. It handles fallback scenarios if necessary and updates the executor
-     * with the given memory information.
+     * @brief Creates an Executor instance based on the provided \memory arguments.
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp:70]: `\memory` -> `memory`

File: src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp [hunk 3]
```diff
@@ -95,104 +67,41 @@ class ExecutorFactory {
...
+     * Depending on the number of available implementations, returns:
+     * - VariableExecutor, if the number of implementations two or more
+     * - simple Executor, if there is only one implementation is available
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp:73]: `simple` -> `Simple`. 
`if there is only one implementation is available` -> `if there is only one available implementation`

File: src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp (hunk 1 above)
Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp]: My bad, removed

File: src/plugins/intel_cpu/src/nodes/fullyconnected.h [hunk 4]
```diff
@@ -85,7 +84,7 @@ class FullyConnected : public Node {
     static const size_t WEIGHTS_ID = 1;
     static const size_t BIAS_ID = 2;
 
-    ExecutorPtr createExecutor();
+    // ExecutorPtr createExecutor();
```
[the remark this answers is not in the document]
Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.h]: Removed

File: src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp (hunk 2 above)
Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp]: I was trying to follow a doxygen syntax for referencing function arguments.
But it does not really matter, probably.
Removed

File: src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp (hunk 3 above)
Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/nodes/executors/executor_factory.hpp]: Corrected

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**